crypto/init.c

```diff
@@ -667,65 +620,59 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
 
 int OPENSSL_atexit(void (*handler)(void))
 {
-    OPENSSL_INIT_STOP *newhand;
-
-#if !defined(OPENSSL_USE_NODELETE)\
-    && !defined(OPENSSL_NO_PINSHARED)
-    {
-# if defined(DSO_WIN32) && !defined(_WIN32_WCE)
-        HMODULE handle = NULL;
-        BOOL ret;
-        union {
-            void *sym;
-            void (*func)(void);
-        } handlersym;
-
-        handlersym.func = handler;
+    if (ossl_no_cleanup != 0)
+        return (0);
 
-        /*
-         * We don't use the DSO route for WIN32 because there is a better
-         * way
-         */
-        ret = GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS
-                                | GET_MODULE_HANDLE_EX_FLAG_PIN,
-                                handlersym.sym, &handle);
-
-        if (!ret)
-            return 0;
-# elif !defined(DSO_NONE)
-        /*
-         * Deliberately leak a reference to the handler. This will force the
-         * library/code containing the handler to remain loaded until we run the
-         * atexit handler. If -znodelete has been used then this is
-         * unnecessary.
-         */
-        DSO *dso = NULL;
-        union {
-            void *sym;
-            void (*func)(void);
-        } handlersym;
-
-        handlersym.func = handler;
+    return ((atexit(handler) == 0) ? 1 : 0);
```

Comment:
No outer parentheses please.

Although it might be OK to keep OPENSSL_atexit() calling atexit(), I wonder if we shouldn't rather add a list of handlers to call in OPENSSL_cleanup() for these.

Also if we keep OPENSSL_atexit() calling atexit() we must absolutely remove all the calls of OPENSSL_atexit() from within the libcrypto and libssl.

Reply:
> Although it might be OK to keep OPENSSL_atexit() calling atexit(), I wonder if we shouldn't rather add a list of handlers to call in OPENSSL_cleanup() for these.

list of handlers is what we have in current (3.3). To be honest I see no benefit in keeping the list of handlers. The handlers we have in 3.3 bring no extra layer of defense against crashes caused by _premature_ `dlclose()`. The defense against this must be provided in libc. Anything else seems like adding complexity for very little benefit.

Reply:
I think we can keep handlers for `OPENSSL_atexit()` around. But for `libssl` we must go for library destructors to keep `shlibloadtest` working. The thing is that `libssl` generally can not rely on `OPENSSL_atexit()` now. The `OPENSSL_atexit()` function is part of `libcrypto` library. The handlers scheduled by `OPENSSL_atexit()` are executed when `libcrypto` library is being unloaded (on `dlclose()`). The `OPENSSL_cleanup()` (invoked to handle `dlclose()`)  may trip over dead memory trying to execute `ssl_library_stop()` from `libssl` library which happens to be unloaded before the `libcrypto`.